File: LC50/lc02.js
class ListNode {
    constructor(x) {
        this.val = x;
        this.next = null;
    }
}


function add_two(n1, n2) {
    var output = new ListNode(0);
    var first = output;
    var plus_one = false;
    var new_val, new_node;
    while (n1 !== null || n2 !== null) {
        new_val = 0;
        if (n1 !== null) {
            new_val += n1.val;
            n1 = n1.next;
        }
        if (n2 !== null) {
            new_val += n2.val;
            n2 = n2.next;
        }
        if (plus_one) {
            new_val += 1;
        }
        if (new_val >= 10) {
            new_val -= 10;
            plus_one = true;
        } else {
            plus_one = false;
        }
        new_node = new ListNode(new_val);
        first.next = new_node;
        first = new_node;
    }
    if (plus_one) {
        first.next = new ListNode(1);
    }
    return output.next;
}


function test() {
    // [2,4,3] +  [5,6,9]
    var n1_3 = new ListNode(3);
    var n1_2 = new ListNode(4);
    var n1_1 = new ListNode(2);
    n1_1.next = n1_2;
    n1_2.next = n1_3;

    var n2_3 = new ListNode(9);
    var n2_2 = new ListNode(6);
    var n2_1 = new ListNode(5);
    n2_1.next = n2_2;
    n2_2.next = n2_3;

    var total = add_two(n1_1, n2_1);
    while (total) {
        console.log(total.val);
        total = total.next;
    }
}


test();

File: LC50/lc02.py
"""
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
"""


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


def add_numbers(number1, number2):
    output = ListNode(0)
    first = output
    plus_one = False
    while number1 or number2:
        new_val = 0
        if number1:
            new_val += number1.val
            number1 = number1.next
        if number2:
            new_val += number2.val
            number2 = number2.next

        if plus_one:
            new_val += 1

        if new_val >= 10:
            new_val -= 10
            plus_one = True
        else:
            plus_one = False

        new_node = ListNode(new_val)
        first.next = new_node
        first = new_node

    if plus_one:
        first.next = ListNode(1)
    return output.next
